File: RGBEngine/src/main/java/engine/graphics/sprites/SpriteSheet.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package engine.graphics.sprites;

import engine.graphics.opengl.Shader;
import engine.graphics.opengl.Texture;
import engine.util.Color;
import engine.util.math.Transformation;
import engine.util.math.Vec2d;

import java.net.URL;
import java.util.HashMap;
import java.util.Map;

import static engine.graphics.opengl.GLObject.bindAll;
import static engine.graphics.sprites.Sprite.SPRITE_VAO;
import static org.lwjgl.opengl.GL11.GL_TRIANGLE_FAN;
import static org.lwjgl.opengl.GL11.glDrawArrays;

/**
 *
 * @author TARS
 */
public class SpriteSheet {

    public static final int SHEET_DEPTH = 32;

    private static final Map<URL, SpriteSheet> SPRITE_SHEET_CACHE = new HashMap();

    public static SpriteSheet load(URL url) {
        if (!SPRITE_SHEET_CACHE.containsKey(url)) {
            SpriteSheet s = new SpriteSheet(url);
            SPRITE_SHEET_CACHE.put(url, s);
        }
        return SPRITE_SHEET_CACHE.get(url);
    }

    public static final Shader SPRITE_SHEET_SHADER = Shader.load(SpriteSheet.class::getResource, "sprite_sheet");

    private final Texture texture;

    private SpriteSheet(URL url) {
        this.texture = Texture.load(url);
    }

    public void draw(Transformation t, int id, Color color) {
        drawTexture(texture, t, id, color);
    }

    public static void drawTexture(Texture texture, Transformation t, int id, Color color) {
        SPRITE_SHEET_SHADER.setMVP(t);
        SPRITE_SHEET_SHADER.setUniform("color", color);
        SPRITE_SHEET_SHADER.setUniform("subCoords", new Vec2d(id % SHEET_DEPTH, 15 - id / SHEET_DEPTH));
        bindAll(texture, SPRITE_SHEET_SHADER, SPRITE_VAO);
        glDrawArrays(GL_TRIANGLE_FAN, 0, 4);
    }

    public int getHeight() {
        return texture.getHeight();
    }

    public int getWidth() {
        return texture.getWidth();
    }
}

File: Cookie/src/main/kotlin/engine/graphics/sprites/SpriteSheet.kt
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package engine.graphics.sprites

import engine.graphics.opengl.BufferObject
import engine.graphics.opengl.GLObject.Companion.bindAll
import engine.graphics.opengl.Shader
import engine.graphics.opengl.Texture
import engine.graphics.opengl.VertexArrayObject
import engine.util.Color
import engine.util.Transformation
import engine.util.Vec2d
import org.lwjgl.opengl.GL11.*
import org.lwjgl.opengl.GL15.GL_ARRAY_BUFFER
import org.lwjgl.opengl.GL20.glEnableVertexAttribArray
import org.lwjgl.opengl.GL20.glVertexAttribPointer
import java.util.*

/**
 *
 * @author TARS
 */
class SpriteSheet private constructor(fileName: String) {

    private val texture: Texture

    val height: Int
        get() = texture.height

    val width: Int
        get() = texture.width

    init {
        this.texture = Texture.load(fileName)
    }

    fun draw(t: Transformation, id: Int, color: Color) {
        drawTexture(texture, t, id, color)
    }

    companion object {

        val SHEET_DEPTH = 32

        private val SPRITE_SHEET_CACHE = HashMap<String, SpriteSheet>()

        fun load(fileName: String): SpriteSheet {
            if (!SPRITE_SHEET_CACHE.containsKey(fileName)) {
                val s = SpriteSheet(fileName)
                SPRITE_SHEET_CACHE.put(fileName, s)
            }
            return SPRITE_SHEET_CACHE.get(fileName)!!
        }

        val SPRITE_SHEET_SHADER = Shader.load("sprite_sheet")

        val SPRITE_SHEET_VAO = VertexArrayObject.createVAO {
            val vbo = BufferObject(GL_ARRAY_BUFFER, floatArrayOf(0.5f, 0.5f, 0f, 1f, 1f, 0.5f, -0.5f, 0f, 1f, 0f, -0.5f, -0.5f, 0f, 0f, 0f, -0.5f, 0.5f, 0f, 0f, 1f))
            glVertexAttribPointer(0, 3, GL_FLOAT, false, 20, 0)
            glEnableVertexAttribArray(0)
            glVertexAttribPointer(1, 2, GL_FLOAT, false, 20, 12)
            glEnableVertexAttribArray(1)
        }

        fun drawTexture(texture: Texture, t: Transformation, id: Int, color: Color) {
            SPRITE_SHEET_SHADER.setMVP(t)
            SPRITE_SHEET_SHADER.setUniform("color", color)
            SPRITE_SHEET_SHADER.setUniform("subCoords", Vec2d((id % SHEET_DEPTH).toDouble(), (15 - id / SHEET_DEPTH).toDouble()))
            bindAll(texture, SPRITE_SHEET_SHADER, SPRITE_SHEET_VAO)
            glDrawArrays(GL_TRIANGLE_FAN, 0, 4)
        }
    }
}
